### PR TITLE
Serve LFS/attachment with http.ServeContent to Support Range-Request

### LINKED ISSUE #16414: [LFS] video player cannot be controlled
The progress bar of the HTML `video` player cannot be dragged, in other words, the player cannot be controlled to play on the specified position. 

To fix this we need support for the `Accept-Ranges: bytes` HTTP header.

### PR DESCRIPTION
**Pull request by marxangels**
The `ServeData` function now can't handle [Range Request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests).
Due to the complexity of this part of HTTP protocol, it is difficult to fully implement it and requires a lot of testing.
The official module `http.ServeContent` has this implementation, and we can replace with it directly.
It can handle `If-Match, If-Unmodified-Since, If-None-Match, If-Modified-Since, If-Range` and more, perfectly!

Thanks for @wxiaoguang 's tips.

### CODE REVIEW

**Comment by wxiaoguang** [comment]
I haven't read through all related code carefully, but I have a feeling that if we call `http.ServeContent` directly, it would break the current behavior.

Currectly, Gitea's `common.ServeData` calls `typesniffer.DetectContentType(buf)` to do deep mime-type detection.

We had better keep the behavior. Since we could get the `ReadSeekCloser`, it's easy to read the a few bytes from beginning to do the `DetectContentType`. And it's better to work on `common.ServeData` directly to introduce `http.ServeContent`, indeed `common.ServeData` does do some specialized work for the response.

**Comment by marxangels** [comment]
According to my experience, the existing mime detection is problematic, some mp3/mp4 files can not be recognized accurately. 
And the `http.ServeContent` found no errors so far.

**Comment by wxiaoguang** [comment]
Yes, there are related issues like:
* https://github.com/go-gitea/gitea/issues/17108
* https://github.com/go-gitea/gitea/issues/13279

Although I also think the `typesniffer.DetectContentType` might not be necessary and buggy for HTTP file serving, the feature was already there. 

If this PR need to change the existing behavior, it need more people @go-gitea/maintainers to take a look at it.

My opinion: I prefer to use extension name to get mime types instead of `typesniffer.DetectContentType` (with some special case handling, eg: detecting text files without extension like README).

**Comment by lunny** [comment]
I think the security headers should be kept. And another problem is `size` is missed. But in fact we can get it.

**Comment by marxangels** [comment]
Updated 🤞

**Comment by lunny** [comment]
`size` should be dropped to support range.

**Comment by lunny** [comment]
The function served attachment/git blob/git lfs file, so I think the name `ServeLargeFile` is not very suitable.

**Comment by marxangels** [comment]
> `size` should be dropped to support range.

`http.ServeContent` uses a seek to the end of the content to determine its size.

**Comment by marxangels** [comment]
> The function served attachment/git blob/git lfs file, so I think the name `ServeLargeFile` is not very suitable.

`git blob` is served by the existing function `common.ServeData` now.
`git blob` doesn't implement a `Seeker` interface, so we should separate it into two functions.

**Comment by zeripath** [comment]
It appears that there's also changes to the mime type detection?

**Comment by marxangels** [comment]
> 

updated

File: modules/typesniffer/typesniffer.go [hunk 1]
```diff
@@ -87,6 +93,26 @@ func DetectContentType(data []byte) SniffedType {
 	return SniffedType{ct}
 }
 
+func DetectContentTypeExtFirst(name string, readSeekerOrBytes interface{}) (SniffedType, error) {
```
Comment (lunny) [inline modules/typesniffer/typesniffer.go]: comment is required by the linter

File: routers/common/repo.go [hunk 2]
```diff
@@ -104,3 +106,34 @@ func ServeData(ctx *context.Context, name string, size int64, reader io.Reader)
 	_, err = io.Copy(ctx.Resp, reader)
 	return err
 }
+
+func ServeLargeFile(ctx *context.Context, name string, time time.Time, reader io.ReadSeeker) error {
+	name = strings.ReplaceAll(name, ",", " ")
+	ctx.Resp.Header().Set("Cache-Control", "public,max-age=86400")
```
Comment (silverwind) [inline routers/common/repo.go]: We shouldn't store in public caches if the gitea instance is private. Also, 1 day seems like a long period, I suggest just specifying `max-age=300` only like GH does.

File: routers/common/repo.go [hunk 3]
```diff
@@ -104,3 +106,34 @@ func ServeData(ctx *context.Context, name string, size int64, reader io.Reader)
...
+		ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf(`inline; filename="%s"`, name))
+		if st.IsSvgImage() {
+			ctx.Resp.Header().Set("Content-Security-Policy", "default-src 'none'; style-src 'unsafe-inline'; sandbox")
+			ctx.Resp.Header().Set("X-Content-Type-Options", "nosniff")
+			ctx.Resp.Header().Set("Content-Type", typesniffer.SvgMimeType)
+		}
+	} else {
+		ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
+		ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, name))
+	}
```
Comment (silverwind) [inline routers/common/repo.go]: This header stuff is at least partially duplicate with `routers/common/repo.go`. Can we introduce a shared function to remove the duplicate parts?

**Comment by marxangels** [comment]
**What's wrong with it?**

image

**Comment by KN4CK3R** [comment]
As you may have noticed it does not fail because of your code. Could you please stop with these commits until someone could have a look?

File: routers/common/repo.go (hunk 2 above)
Comment (marxangels) [reply routers/common/repo.go]: need help

> As you may have noticed it does not fail because of your code. Could you please stop with these commits until someone could have a look?

Okay~ the build check is too slow to continue.

File: routers/common/repo.go [hunk 4]
```diff
@@ -21,7 +22,56 @@ import (
 	"code.gitea.io/gitea/modules/util"
 )
 
-// ServeBlob download a git.Blob
+func setCommonHeaders(ctx *context.Context, name string, data interface{}) error {
+	// Google Chrome dislike commas in filenames, so let's change it to a space
+	name = strings.ReplaceAll(name, ",", " ")
```
Comment (Gusted) [inline routers/common/repo.go]: Ditto from the other PR:



I don't think this is needed actually, from what I remember about this weird bug, if you quote the filename(as being done in this function) chromium won't error about it.

Comment (marxangels) [reply routers/common/repo.go]: That's strange. Is it dependent on different chromium versions?

Comment (wxiaoguang) [reply routers/common/repo.go]: It's a historical problem, I believe modern browsers (and modern frameworks) can handle it correctly, and we do not need hacky methods.

https://stackoverflow.com/questions/93551/how-to-encode-the-filename-parameter-of-content-disposition-header-in-http

File: integrations/download_test.go [hunk 5]
```diff
@@ -73,19 +74,22 @@ func TestDownloadRawTextFileWithoutMimeTypeMapping(t *testing.T) {
 	req := NewRequest(t, "GET", "/user2/repo2/raw/branch/master/test.xml")
 	resp := session.MakeRequest(t, req, http.StatusOK)
 
-	assert.Equal(t, "text/plain; charset=utf-8", resp.HeaderMap.Get("Content-Type"))
+	assert.Equal(t, "text/xml; charset=utf-8", resp.HeaderMap.Get("Content-Type"))
```
Comment (6543) [inline integrations/download_test.go]: -> RawTextFileWithoutMimeTypeMapping this test is intended to not not find a specific mime type

Comment (marxangels) [reply integrations/download_test.go]: Yeah, but the type-detection is now taken care by `mime.TypeByExtension`. And the default mime of XML file is `text/xml`.

Comment (marxangels) [reply integrations/download_test.go]: The type-detection of text files has known logic confusion now. we may fix it after this PR.
There are too many modifications in this PR.

Comment (marxangels) [reply integrations/download_test.go]: To test a specific mime type not found, we can use any unknown extension, such as `.foo` `.xyz`

**Comment by marxangels** [comment]
Updated.
Please let me know if any improvement is needed.
Thank you all so much!

File: routers/common/repo.go [hunk 6]
```diff
@@ -21,7 +22,63 @@ import (
...
+
+	// Fixme: ctx.Repo.Repository.IsPrivate is nil
+	// if ctx.Repo.Repository.IsPrivate {
+	// 	ctx.Resp.Header().Set("Cache-Control", "private, max-age=300")
+	// } else {
+	// 	ctx.Resp.Header().Set("Cache-Control", "public, max-age=86400")
+	// }
+	ctx.Resp.Header().Set("Cache-Control", "public, max-age=86400")
```
Comment (silverwind) [inline routers/common/repo.go]: Already commented on this, 86400 is far too long. Would suggest default 300, ideally also exposed in a config variable, very similar to `STATIC_CACHE_TIME`, maybe call it `CONTENT_CACHE_TIME`.

Comment (marxangels) [reply routers/common/repo.go]: Updated

File: routers/common/repo.go [hunk 7]
```diff
@@ -21,7 +22,63 @@ import (
 	"code.gitea.io/gitea/modules/util"
 )
 
-// ServeBlob download a git.Blob
+func setCommonHeaders(ctx *context.Context, name string, data interface{}) error {
+	// Google Chrome dislike commas in filenames, so let's change it to a space
+	name = strings.ReplaceAll(name, ",", " ")
+
+	// Fixme: ctx.Repo.Repository.IsPrivate is nil
```
Comment (6543) [inline routers/common/repo.go]: where is this issue coming from?

Comment (marxangels) [reply routers/common/repo.go]: https://github.com/go-gitea/gitea/commit/1bfb0a24d843e10d6d95c4319a84980485e584ed#diff-6da01db79cc03b39079c8ef80e36c1c96cb83b7e0c0c65086d5f84ba79c6ae44R88-R89

File: integrations/download_test.go [hunk 8]
```diff
@@ -70,24 +71,26 @@ func TestDownloadRawTextFileWithoutMimeTypeMapping(t *testing.T) {
 
 	session := loginUser(t, "user2")
 
-	req := NewRequest(t, "GET", "/user2/repo2/raw/branch/master/test.xml")
+	req := NewRequest(t, "GET", "/user2/repo2/raw/branch/master/bin.foo")
```
Comment (silverwind) [inline integrations/download_test.go:101]: Could you just add additional test cases instead of modifying existing ones?

Comment (marxangels) [reply integrations/download_test.go:101]: Good to know.

Comment (silverwind) [reply integrations/download_test.go:101]: Yeah, it's generally good to only add tests, never remove them. That way you can be sure to not regress those existing tests.

File: routers/common/repo.go [hunk 9]
```diff
@@ -21,7 +22,57 @@ import (
...
+	// reset the offset to the start of served file
+	if seeker, ok := data.(io.ReadSeeker); ok {
+		_, _ = seeker.Seek(0, io.SeekStart)
+	}
+
+	if st.IsText() || ctx.FormBool("render") {
+		var cs string
+		var err error
+		if reader, ok := data.(io.ReadSeeker); ok {
+			cs, err = charset.DetectEncodingFromReader(reader)
```
Comment (6543) [inline routers/common/repo.go]: do we need a seeker reader explicit for `http.ServeContent` - else change the detection to read in a buffer and use multi reader would also be a good option

File: modules/typesniffer/typesniffer.go [hunk 10]
```diff
@@ -78,15 +85,35 @@ func DetectContentType(data []byte) SniffedType {
 		data = data[:sniffLen]
 	}
 
-	if (strings.Contains(ct, "text/plain") || strings.Contains(ct, "text/html")) && svgTagRegex.Match(data) ||
-		strings.Contains(ct, "text/xml") && svgTagInXMLRegex.Match(data) {
+	if (strings.HasPrefix(ct, "text/plain") || strings.HasPrefix(ct, "text/html")) && svgTagRegex.Match(data) ||
```
Comment (lunny) [inline modules/typesniffer/typesniffer.go:92]: Why not keep `Contains`?

**Comment by LJFan** [comment]
I need this feature badly. Can anyone merge it into the main branch?

File: integrations/download_test.go [hunk 11]
```diff
@@ -78,8 +79,10 @@ func TestDownloadRawTextFileWithoutMimeTypeMapping(t *testing.T) {
 
 func TestDownloadRawTextFileWithMimeTypeMapping(t *testing.T) {
 	defer prepareTestEnv(t)()
+
```
Comment (6543) [inline integrations/download_test.go]:
```suggestion
```

File: integrations/download_test.go [hunk 12]
```diff
@@ -78,8 +79,10 @@ func TestDownloadRawTextFileWithoutMimeTypeMapping(t *testing.T) {
...
 	setting.MimeTypeMap.Map[".xml"] = "text/xml"
 	setting.MimeTypeMap.Enabled = true
+	_ = mime.AddExtensionType(".xml", "text/xml")
```
Comment (6543) [inline integrations/download_test.go]:
```suggestion
```

File: modules/setting/mime_type_map.go [hunk 13]
```diff
@@ -21,6 +26,10 @@ func newMimeTypeMap() {
 	m := make(map[string]string, len(keys))
 	for _, key := range keys {
 		m[strings.ToLower(key.Name())] = key.Value()
+		err := mime.AddExtensionType(key.Name(), key.Value())
```
Comment (6543) [inline modules/setting/mime_type_map.go:29]: please dont add unrelated changes in here

**Comment by 6543** [comment]
> I need this feature badly. Can anyone merge it into the main branch?

as soon as it's ready ...

**Comment by 6543** [comment]
@marxangels whats going on?

**Comment by silverwind** [comment]
Maybe he lost interest, I think we should re-raise the PR in a cleaned up form. It has some valuable stuff in it.

**Comment by 6543** [comment]
-> #20480

I put it on my list of WIPs ...

**Comment by marxangels** [comment]
No line of code is irrelevant, but it's hard for me to explain it clearly. Too much code involved.
